[team11][iOS] - 사용자 위치 정보 확인 및 CustomCallendar 작성 (2주차 화요일 PR)

wnsxor1993 opened the pull request:
# [team11][iOS] - CLLocationManager를 이용한 사용자 위치 정보 확인 및 CustomCallendar 작성

## 💡 Issue
1. CLLocationManager를 사용하여 사용자 위치 정보 접근 권한 획득 및 이를 통한 정보 데이터 확인
2. 서칭 뷰 관련 로직 분리 및 내부 CollectionView 하나만 사용하여 layout, dataSource 스위칭 구현
3. CustomCalendar 구현
   

## 📝 Tasks

## Task1

#### 🔹 작업
- BrowseViewController에 구현한 서칭 뷰와 관련하여 collectionViewLayout 로직 분리하여 관리
- BrowseViewController에 구현된 CollectionView를 하나로 통합하고 layout과 dataSource를 스위칭하여 사용자 작업에 맞춰 view 변화 구현

#### 🔹 고민과 해결
- CollectionView를 두 개나 사용하고 사용자 작업 환경에 맞춰 view를 add하거나 remove 해주는 것이 비효율적이라 생각됨.
이에 CollectionView는 하나로 두고, 내부의 프로퍼티에 따라 분기를 두어 사용자 작업 환경에 따라 layout과 dataSource를 스위칭하여 view를 변화시켜주는 방식으로 로직 변경.
스위칭해주는 메서드의 호출 위치나 reloadData를 호출해주는 위치에 대한 고민이 있었음.

- [고민사항]
    MKLocalSearchCompleter를 활용하여 위치를 서칭할 때, 해외의 스팟이 나오거나 각 도시 별로 출력되는 정보의 질이나 양의 차이가 심하여 중구난방의 느낌이 강합니다.
    이를 고치고자 다양한 자료를 보면서 여러 방안을 시도해보았으나 별다른 차이가 없거나, MapView를 통해 현재 사용자의 위치를 기반으로 찾는 방식이 대다수라 적절하게 수정하기 어려웠습니다.
    혹시 이와 관련하여 어떤 방식으로 접근하는 것이 좋을 지 여쭤보고자 합니다.


### Task2

#### 🔹 작업
- CLLocationManager를 통해서 어플을 사용 시에 사용자 위치 정보에 대한 접근 권한을 요청하도록 구현
- 접근 권한에 맞춰 사용자 위치 정보를 받아오고 이를 프로퍼티에 담아주는 작업을 구현

#### 🔹 고민과 해결 
- 사용자 위치 정보 접근 권한을 요청하고자 Delegate에 있는 메서드를 활용하였으나 전혀 호출이 되지 않았습니다. 그러던 중, 해당 메서드들이 status가 변경되었을 때 호출되는 것을 확인하였고 이에 홈뷰가 불려올 때 현재 locationManager의 상태에 따라 접근을 요청하거나 위치 정보에 접근하는 메서드들을 호출하도록 분기 처리하는 로직을 작성하였고, 정상적으로 작동하는 것을 확인하였습니다.


- [고민 사항]
    사용자가 위치 정보 접근에 대하여 denied 했을 때, 설정을 통해 다시 설정하도록 권유하는 로직이 있을 것 같은데 이와 관련한 내용이 잘 나오지 않아 이 부분에 대한 질문을 드리고자 합니다.

### Task3

#### 🔹 작업
- CustomCalendar를 구현하기 위한 각 달의 날짜 구하는 로직 구현
- 현재 날짜 이전의 날짜의 Cell들은 터치가 되지않고, `-` 표시가 있는 회색 Label.text를 그리는 기능 구현

#### 🔹 고민과 해결 
- Calendar에 해당하는 CollectionView의 각 Section(월단위)의 HeaderView의 size를 결정해주는 메소드 구현 중에, Label의 text가 1줄일 때를 기준으로 해당 HeaderView의 높이가 유동적으로 결정되는 로직을 구현하려 했습니다.
    해당 HeaderView의 Label에는 `숫자 + 년 + 숫자 + 월` 이라는 text가 들어갈 예정이라, `년, 월`에 해당하는 글자의 높이만큼의 height를 가지면 된다고 판단하여 `string: NSString = "년월"`을 기준으로 HeaderView의 높이를 결정하도록 했습니다. 
- [고민 사항] Date 타입을 이용해 날짜를 나타내고, 해당 값에 month, day를 빼고 더해서 원하는 날짜를 찾으려고 했습니다. 그런데 Date의 값을 변화시킬때마다 원하는 날짜가 아닌  날짜를 저장하는 문제가 있었습니다. DateFormatter를 이용한다면 또 원하는 날짜가 표현되어, 혹시 Date 타입을 View 등에 나타내거나 다른 VC로 전달해주려고 할 때는 DateFormatter를 통한 변형이 필수적인지 궁금합니다!
- [고민 사항] 저번주 수요일 PR에서, CollectionView의 HeaderView의 높이가 정해져있기 때문에 label이 너무 길어질 경우 원하는 모습대로 뷰가 그려지지 않는다는 말씀을 들었는데, 이를 해결하려고 하는 도중 Header에 들어갈 문자를 Layout을 그려주는 객체가 알게 되는게 맞는지에 대한 의문이 들었습니다. Layout을 받아온 후 CollectionView를 그리는 흐름으로 이해하여, HeaderView에 들어갈 문자를 미리 알아내서 layout을 그리는 객체로 넘겨주는게 맞는걸까요? 
    
## 📱 실행화면
![SS 2022-05-31 PM 05 13 47](https://user-images.githubusercontent.com/92504186/171125912-8b2ee27e-0699-41ab-a7d5-ba22515ccb72.gif)

Hansolkkim committed:
refactor: DateDate 구조체 이름을 DateConverter로 변경 (wnsxor1993/airbnb_team11/#22)

Date 표기방법을 변환하는 객체로써의 이름을 가지도록, DateConverter라는 이름으로 구조체 이름 변경

Hansolkkim committed:
refactor: 분기문의 조건을 상수로 정의해서 어떤 조건인지 명시적으로 나타내도록 수정 (wnsxor1993/aribnb_team11/#22)

wnsxor1993 committed:
refactor: BrowseViewCollectionLayout 로직 분리

wnsxor1993 committed:
refactor: 분기 처리를 통한 collectionLayout, dataSource 분리

wnsxor1993 committed:
feat: 지우기 버튼 추가

wnsxor1993 committed:
refactor: collectionDataSource image 변경

Hansolkkim committed:
chore: "HorizonCalendar" pod파일 제거 및 관련 파일 제거 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: "HorizonCalendar" pod파일 제거 및 관련 파일 제거 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: Custom Calendar 구현 (아직 현재 시점의 달만 보여지도록 구현) (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: Custom Calendar 구현 (아직 현재 시점의 달만 보여지도록 구현) (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 달력의 날짜가 선택됐을 때의 흐름 구현 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 달력의 날짜가 선택됐을 때의 흐름 구현 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 오늘 이전의 날짜는 선택이 되지 않는 기능 추가 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 오늘 이전의 날짜는 선택이 되지 않는 기능 추가 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: 사용하지 않는 코드 제거 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: 사용하지 않는 코드 제거 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 달력 뷰의 Header로 사용할 View 구현 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 달력 뷰의 Header로 사용할 View 구현 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: 사용하지 않는 메소드 정리 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: 사용하지 않는 메소드 정리 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 오늘 이전의 날들의 Label에 줄이 그어지고 textColor가 다르도록 표현하는 기능 추가 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 오늘 이전의 날들의 Label에 줄이 그어지고 textColor가 다르도록 표현하는 기능 추가 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: 사용하지 않는 코드 제거 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: 사용하지 않는 코드 제거 (wnsxor1993/airbnb_team11#23)

wnsxor1993 committed:
feat: CLLocationManager로 위치정보 받아오기 및 위치 권한 요청 구현

wnsxor1993 committed:
feat: User의 현 위치 받아오기 기능 구현

Hansolkkim committed:
Merge pull request #27 from wnsxor1993/refactor/BrowseView

[iOS] Refactor: Browse 화면 리팩토링

wnsxor1993 committed:
Merge pull request #26 from wnsxor1993/refactor/secondReview

[iOS] Refactor: 두번째 리뷰 사항 수정

wnsxor1993 committed:
refactor: 권한 요청 관련 로직 수정

Hansolkkim committed:
fix: Cell을 재사용할 때 attributedText 속성이 제거되지 않던 문제 해결 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
fix: Cell을 재사용할 때 attributedText 속성이 제거되지 않던 문제 해결 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 달력을 현재로부터 12개월 후까지 표시하도록 뷰 수정 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 달력을 현재로부터 12개월 후까지 표시하도록 뷰 수정 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
Merge pull request #30 from wnsxor1993/feat/CoreLocation

[iOS] 사용자 위치 권한 요청 및 위치 데이터 할당 기능 추가

Hansolkkim committed:
feat: CalenderView의 collectionView HeaderView를 설정해주기 위한 수정 (wnsxor1993/airbnb_team11#23)

기존에 그냥 UIView로 사용하던 CalendarCollectionHeaderView를 해당 CollectionView의 HeaderView로 사용하기 위한 수정

Hansolkkim committed:
feat: CalenderView의 collectionView HeaderView를 설정해주기 위한 수정 (wnsxor1993/airbnb_team11#23)

기존에 그냥 UIView로 사용하던 CalendarCollectionHeaderView를 해당 CollectionView의 HeaderView로 사용하기 위한 수정

Hansolkkim committed:
feat: CalendarView의 collectionView의 headerView 높이를 계산해주는 로직 추가 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: CalendarView의 collectionView의 headerView 높이를 계산해주는 로직 추가 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 선택된 두 날짜를 findAccomodationView의 tableView의 두번째 cell에 표시되는 기능 추가 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
feat: 선택된 두 날짜를 findAccomodationView의 tableView의 두번째 cell에 표시되는 기능 추가 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: 날짜 표현 코드 수정 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
chore: 날짜 표현 코드 수정 (wnsxor1993/airbnb_team11#23)

Hansolkkim committed:
Merge branch 'refactor/secondBrowseView' of https://github.com/wnsxor1993/airbnb_team11 into refactor/secondBrowseView

# Conflicts:
#	iOS/Airbnb_iOS/Airbnb_iOS.xcodeproj/project.pbxproj

wnsxor1993 committed:
Merge pull request #31 from wnsxor1993/refactor/secondBrowseView

[iOS] Custom Calendar 구현 + refactoring

wnsxor1993 committed:
refactor: print 구문 삭제 및 검색 출력 코드 수정

wnsxor1993 committed:
Merge pull request #33 from wnsxor1993/refactor/HomeViewToCalendarView

[iOS] Refactor: 일부 구문 수정

Hansolkkim committed:
fix: 체크인/체크아웃에 표시되는 예약날짜가 UTC 기준으로 표시되는 이슈 해결 (wnsxor1993/airbnb_team11#32)

DateConverter 객체 수정을 통해 해당 문제 해결

wnsxor1993 committed:
refactor: searchViewController dismiss 구문 추가

Hansolkkim committed:
fix: FindAccomodationVC에서 이전 VC로 돌아갔을 때 Toolbar가 사라지지 않는 이슈 해결 (wnsxor1993/airbnb_team11#32)

viewWillAppear() 메소드와 viewWillDisappear() 메소드를 overriding하여 해당 이슈 해결

Hansolkkim committed:
Merge pull request wnsxor1993/airbnb_team11#38 from wnsxor1993/refactor/HomeViewToCalendarView

[iOS] Calendar 날짜 수정 및 SearchVC dismiss 추가

mienne commented:
뷰 디버깅에서 BrowsingSpotCell 에서 Position is ambiguous for - 메세지가 뜨고 있어요. 사용자에게 뷰가 깨져서 보이는 것은 아니지만, 뷰 성능상 문제가 될 수 있습니다. 앱 작업할 때 뷰 디버거로 경고가 뜨는지 확인해주세요!

[참고: Solving ambiguous constraints without rerunning your app](https://mar.codes/blog/constraints)

mienne commented:
* MKLocalSearchCompleter를 활용하여 위치를 서칭할 때, 해외의 스팟이 나오거나 각 도시 별로 출력되는 정보의 질이나 양의 차이가 심하여 중구난방의 느낌이 강합니다.
이를 고치고자 다양한 자료를 보면서 여러 방안을 시도해보았으나 별다른 차이가 없거나, MapView를 통해 현재 사용자의 위치를 기반으로 찾는 방식이 대다수라 적절하게 수정하기 어려웠습니다.
혹시 이와 관련하여 어떤 방식으로 접근하는 것이 좋을 지 여쭤보고자 합니다.

> 아쉽게도 앱에서 위치 관련된 데이터 보정하는 것은 한계가 있습니다ㅠㅠ 일반적으로 사용자 위/경도 정보를 서버로 보내고, 받은 정보를 적절하게 데이터를 가공하여 앱으로 응답을 보내줍니다. 그나마 앱에서 더 정밀한 정보를 얻고 싶다면 결국 TMap이나 [구글맵]( https://mapsplatform.google.com/pricing/?hl=ko) 이용하는 방법인데 돈을 내야하기 때문에...😇..... 한국으로 범위를 좁힌다면 차선책으로 네이버맵을 사용하는 방법도 있긴 합니다🤔

mienne commented:
* 사용자가 위치 정보 접근에 대하여 denied 했을 때, 설정을 통해 다시 설정하도록 권유하는 로직이 있을 것 같은데 이와 관련한 내용이 잘 나오지 않아 이 부분에 대한 질문을 드리고자 합니다.

> 사용자 위치 정보 접근 권한을 선택할 때 UserDefault에다가 플래그 값 여부로 저장하고, UserDefault 저장한 값에 따라 권한을 다시 얻을 수 있도록 호출하면 되지 않을까 생각합니다. 다만, 다시 설정하도록 권유하는 로직은 앱 실행? 화면에 접근할 때 이런 부분은 정하면 될 것 같습니다. 위치 기반 서비스로 대표적으로 당근마켓, 배달의 민족이 있는데 사용자의 위치 정보 접근 권한이 off 되어 있을 때 어떤 시점에서 재설정 묻는지 참고하면 도움이 될거 같습니다🙂

이 부분은 제가 제대로 이해하고 답변 드린게 맞는지 확인 부탁드릴게요!

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/Model/UseCase/CalendarViewControllerUseCase.swift:
```diff
@@ -0,0 +1,60 @@
+    }
+}
+
+protocol CalendarViewControllerUseCaseDelegate: AnyObject {
```
`CalendarViewControllerUseCaseDelegate` 파일 하단보다 상단에 있다면 파일 열자마자 Delegate가 가지고 있는 인터페이스를 확인할 수 있습니다👀

mienne replied:
코드 배치를 어떻게 하냐에 따라 가독성이 달라질 수 있습니다. 내가 만약 이 코드에 대해서 전혀 모르는 상태에서 아래 3가지 케이스로 배치한다면 어떤 케이스에서 더 빨리 접근할 수 있을지 생각해보시면 좋을거 같아요🤔

1. CalendarViewControllerUseCaseDelegate 파일 하단에 있는 경우
2. CalendarViewControllerUseCaseDelegate 파일 상단에 있는 경우
3. 사용하는 도메인 내에서 Protocols 그룹을 만들고 인터페이스를 모아놓는 경우

Hansolkkim replied:
가독성 측면에서는 말씀하신 케이스 중 2번째 케이스가 제일 좋을 것 같습니다!
해당 케이스처럼 수정하고, 폴더도 도메인별로 나누어서 조금 더 가독성 좋은 프로젝트가 될 수 있도록 수정해보겠습니다!

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/Utility/Converter/DateConverter.swift:
```diff
@@ -0,0 +1,24 @@
+        return "\(dateFormatter.string(from: startDate)) - \(dateFormatter.string(from: endDate))"
+    }
+
+    init(dateRange: ClosedRange<Date>) {
```
유틸성 객체인 DateConverter 객체 생성하고, 필요하지 않을 때 메모리 해제하는 과정이 계속 반복해야 하는 작업은 아닌거 같아요. 어느 화면에서 필요한 시점 호출할 수 있도록 메모리 계속 올려놓는게 적절하지 않을까 생각합니다.


```swift
struct DateConverter {

    private static let dateFormatter = DateFormatter()

    static func convert(from fromDate: date, to toDate: Date) {
        dateFormatter.dateFormat = "M월 d일"
        dateFormatter.timeZone = .autoupdatingCurrent
        return "\(dateFormatter.string(from: fromDate)) - \(dateFormatter.string(from: toDate))"
    }
}
```

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/View/FindAccomodationView/CalendarCollectionHeaderView.swift:
```diff
@@ -0,0 +1,56 @@
+    private let monthLabel: UILabel = {
+        let label = UILabel()
+        label.translatesAutoresizingMaskIntoConstraints = false
+        label.font = .systemFont(ofSize: 16, weight: .init(rawValue: 700))
```
예전에 리뷰로 질문했던 부분인데, 피그마로 확인하니 weight 값이 숫자로 나오더라구요. 그래서 코드가 왜 이렇게 되었나 이제서야 이해합니다😭

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/View/FindAccomodationView/CalendarCollectionHeaderView.swift:
```diff
@@ -0,0 +1,56 @@
+        return label
+    }()
+
+    private lazy var dateFormatter: DateFormatter = {
```
DateFormatter 관련된 로직은 DateConverter 이미 그런 역할을 하고 있으니 그 쪽으로 이동하는 것이 적절한거 같아요!

Hansolkkim replied:
말씀하신 것처럼, DateConverter와 비슷한 역할을 하는 `dateFormatter`라는 프로퍼티를 FincAccomodationVC, CalendarVC, CalendarHeaderView에서 각각 가지고 있을 필요가 없다고 생각되네요..
위에서 말씀하신 static function 사용하는 방법으로 DateFormatter 관련 로직을 DateConverter에서 가지도록 수정하곘습니다!

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/ViewController/TabBarItemVC/SearchViewController.swift:
```diff
@@ -34,10 +35,54 @@ class SearchViewController: UIViewController {
     }
+    
+    func setLocationAccess() {
+        guard CLLocationManager.locationServicesEnabled() else { return }
```
1. SearchViewControlle에서 CLLocationManager 권한 체크하는 로직을 알고 있어야 하는 부분이 적절하지 않은거 같아요. SearchViewController에서 다른 객체를 통해서 사용자 위치 권한을 설정할 수 있도록 실행해줘 라는 명령어만 보낼 수 있지 않을까요?
2. 위 답변에 이어서 CLLocationManager 가지고 있는 LocationManager 싱글톤 객체로 만들면 어떨까요?
3. 만약 사용자가 위치 접근 권한 허용하지 않았다면, 앱이 다시 실행할 때 권한을 다시 물어볼 수 있는 메서드도 필요할거 같아요!

wnsxor1993 replied:
CLLocationManager에 대한 객체를 따로 생성한 뒤, access를 설정해주는 메서드를 분리하였습니다. 이때 해당 메서드에서는 enum 타입을 리턴하고 이에 따라 VC에서는 해당 리턴 값이 사용자가 권한을 거부한 denied에 해당할 경우에는 거부 버튼을 눌렀을 때와 이후 앱을 재실행할 때마다 권한 설정이 필요하므로 애플의 설정창에 들어가서 변경을 부탁하는 Alert를 띄어주도록 하였습니다.
또한 해당 Alert는 해당 창을 단순히 끄는 닫기 버튼과 바로 애플의 설정창으로 이동시켜주는 버튼을 각각 구현하였습니다.

말씀주신 싱글톤 객체에 대해서는 고민해보았으나 여러곳에서 자주 접근할 만한 요소가 아니라는 생각이 들어 일단은 일반적인 객체로 작성하였으니 확인 부탁드리겠습니다.

mienne reviewed (approved):
메일을 늦게 확인해서 이제서야 리뷰 남깁니다.  리뷰 확인하면서 질문해주신 고민사항에 대해서 추가적인 답변 남긴 것 같아요. 확인 부탁드릴게요. 답변이 애매한 부분은 언제든지 슬랙 디엠 주시면 답변 드리도록 하겠습니다!

수고하셨습니다👍

Hansolkkim committed:
refactor: DateConverter 수정 (wnsxor1993/airbnb_team11#39)

DateConverter객체를 메모리에서 계속 생성, 해제하지 않도록, convert를 위한 메소드들을 static function으로 수정
Date를 원하는 모습으로 수정하길 원하는 모든 객체에서 DateFormatter를 따로 소유하지 않고 DateConverter를 이용하도록 수정

Hansolkkim committed:
style: 가독성을 위해 delegate protocol 의 파일 내 위치 변경 (wnsxor1993/airbnb_team11#39)

wnsxor1993 committed:
refactor: BrowsingSpotCell autolayout 수정

wnsxor1993 committed:
feat: 사용자 위치 권한 denied시, 설정 이동 alert 추가

Hansolkkim committed:
rename: 도메인별로 폴더 분리 및 SearchViewController -> HomeViewController 이름 수정 (wnsxor1993/airbnb_team11#39)

wnsxor1993 committed:
refactor: LocationManager 객체 생성을 통해 로직 분리

wnsxor1993 commented:
> * MKLocalSearchCompleter를 활용하여 위치를 서칭할 때, 해외의 스팟이 나오거나 각 도시 별로 출력되는 정보의 질이나 양의 차이가 심하여 중구난방의 느낌이 강합니다.
>   이를 고치고자 다양한 자료를 보면서 여러 방안을 시도해보았으나 별다른 차이가 없거나, MapView를 통해 현재 사용자의 위치를 기반으로 찾는 방식이 대다수라 적절하게 수정하기 어려웠습니다.
>   혹시 이와 관련하여 어떤 방식으로 접근하는 것이 좋을 지 여쭤보고자 합니다.
> 
> > 아쉽게도 앱에서 위치 관련된 데이터 보정하는 것은 한계가 있습니다ㅠㅠ 일반적으로 사용자 위/경도 정보를 서버로 보내고, 받은 정보를 적절하게 데이터를 가공하여 앱으로 응답을 보내줍니다. 그나마 앱에서 더 정밀한 정보를 얻고 싶다면 결국 TMap이나 [구글맵](https://mapsplatform.google.com/pricing/?hl=ko) 이용하는 방법인데 돈을 내야하기 때문에...😇..... 한국으로 범위를 좁힌다면 차선책으로 네이버맵을 사용하는 방법도 있긴 합니다🤔
>
>>> 답변 감사합니다! 아무래도 쉽지 않을 거라 생각이 들었지만 너무 서칭 리스트가 두서없는 느낌이라 조금이라도 깔끔하게 다듬어보려고 찾아봤는데 역시나 정보가 잘 없더군요ㅠㅠ

wnsxor1993 commented:
> * 사용자가 위치 정보 접근에 대하여 denied 했을 때, 설정을 통해 다시 설정하도록 권유하는 로직이 있을 것 같은데 이와 관련한 내용이 잘 나오지 않아 이 부분에 대한 질문을 드리고자 합니다.
> 
> > 사용자 위치 정보 접근 권한을 선택할 때 UserDefault에다가 플래그 값 여부로 저장하고, UserDefault 저장한 값에 따라 권한을 다시 얻을 수 있도록 호출하면 되지 않을까 생각합니다. 다만, 다시 설정하도록 권유하는 로직은 앱 실행? 화면에 접근할 때 이런 부분은 정하면 될 것 같습니다. 위치 기반 서비스로 대표적으로 당근마켓, 배달의 민족이 있는데 사용자의 위치 정보 접근 권한이 off 되어 있을 때 어떤 시점에서 재설정 묻는지 참고하면 도움이 될거 같습니다🙂
> 
> 이 부분은 제가 제대로 이해하고 답변 드린게 맞는지 확인 부탁드릴게요!

권한 설정이 denied가 된 상태에서는 requestWhenInUseAuthorization 메서드가 더이상 호출이 되지 않았으며, 이에 애플 설정창으로 들어가는 링크를 주고 해당 위치에서 권한 조정을 부탁하는 방법을 여쭤본 것이었습니다.
현재 말씀하셨던 리팩토링 부분을 하던 중 관련 방법을 서칭하여서 적용해본 상태이므로 접근 권한 요청의 시기나, denied 시의 해결 방법이 적절한 지 다시금 확인을 부탁드리고자 합니다.
감사합니다😁😁😁

Hansolkkim committed:
Merge pull request wnsxor1993/airbnb_team11#40 from wnsxor1993/refactor/iOS-thirdReview

[iOS] 3차 PR 리뷰 관련 리팩토링

Hansolkkim commented:
안녕하세요 두두! 
말씀하신 부분들에 대해서 수정 진행해봤습니다!
확인부탁드리겠습니다ㅎㅎ

wnsxor1993 committed:
fix: 뷰 이동 후 스크롤링이 안 되는 문제 해결

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS.xcodeproj/project.pbxproj:
```diff
@@ -292,25 +415,6 @@
-			);
-			path = Container;
-			sourceTree = "<group>";
-		};
```
폴더 위치랑 이름이 변경점이 많네요! 구조 많이 바꾸셨네요!

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/Home/Controller/HomewViewController.swift:
```diff
@@ -54,37 +54,11 @@ private extension HomewViewController {
-                return
-            }
-        }
+        self.locationManager.locationManager.delegate = self
```
👍

mienne commented:
> > * 사용자가 위치 정보 접근에 대하여 denied 했을 때, 설정을 통해 다시 설정하도록 권유하는 로직이 있을 것 같은데 이와 관련한 내용이 잘 나오지 않아 이 부분에 대한 질문을 드리고자 합니다.
> > 
> > > 사용자 위치 정보 접근 권한을 선택할 때 UserDefault에다가 플래그 값 여부로 저장하고, UserDefault 저장한 값에 따라 권한을 다시 얻을 수 있도록 호출하면 되지 않을까 생각합니다. 다만, 다시 설정하도록 권유하는 로직은 앱 실행? 화면에 접근할 때 이런 부분은 정하면 될 것 같습니다. 위치 기반 서비스로 대표적으로 당근마켓, 배달의 민족이 있는데 사용자의 위치 정보 접근 권한이 off 되어 있을 때 어떤 시점에서 재설정 묻는지 참고하면 도움이 될거 같습니다🙂
> > 
> > 
> > 이 부분은 제가 제대로 이해하고 답변 드린게 맞는지 확인 부탁드릴게요!
> 
> 권한 설정이 denied가 된 상태에서는 requestWhenInUseAuthorization 메서드가 더이상 호출이 되지 않았으며, 이에 애플 설정창으로 들어가는 링크를 주고 해당 위치에서 권한 조정을 부탁하는 방법을 여쭤본 것이었습니다. 현재 말씀하셨던 리팩토링 부분을 하던 중 관련 방법을 서칭하여서 적용해본 상태이므로 접근 권한 요청의 시기나, denied 시의 해결 방법이 적절한 지 다시금 확인을 부탁드리고자 합니다. 감사합니다😁😁😁

역시 제가 잘못 이해한 부분이 있었네요. 찾으셔서 적용 잘하신거 같아요👍 `denied` 상태일 때 사용자에게 설정으로 다시 요청하는 방법도 처리 방법 중 하나이긴 해요!

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/Home/Controller/HomewViewController.swift:
```diff
@@ -0,0 +1,127 @@
+    }
+    
+    func alertLocationAccessNeeded(isDenied: LocationManager.AceessCase) {
+        var settingsAppURL: URL?
```
기본값을 주고 아래처럼 else문 제거하는 방법이 있습니다🙂

```suggestion
        var settingsAppURL: URL? = URL(string: UIApplication.openSettingsURLString)
        
        if #available(iOS 15.4, *) {
            settingsAppURL = URL(string: UIApplicationOpenNotificationSettingsURLString)
        }
```

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/Home/Controller/HomewViewController.swift:
```diff
@@ -0,0 +1,127 @@
+    
+    func setLocationAccess() {
+        self.locationManager.locationManager.delegate = self
+        self.alertLocationAccessNeeded(isDenied: self.locationManager.setLocationAccess())
```
```suggestion
        self.alertLocationAccessIfNeeded(isDenied: self.locationManager.setLocationAccess())
```

mienne commented on iOS/Airbnb_iOS/Airbnb_iOS/Home/Model/UseCase/LocationManager.swift:
```diff
@@ -0,0 +1,54 @@
+
+final class LocationManager {
+    
+    let locationManager = CLLocationManager()
```
위치 관련해서 말씀 드리지만 지금 코드상으로 문제가 되지 않지만, `CLLocationManagerDelegate` 위치 서비스가 시작한 곳의 스레드에서 동작하기 때문에 간혹 백그라운드로 갔다가 UIAlert 같은 UIKit 사용하게 되면 메인스레드가 아닌 곳에서 동작하기 때문에 크래시가 날 수 있어요. 그래서 보통 권한 설정 후, Alert창 띄울 때 메인 스레드에서 동작할 수 있도록 DispatchQueue로 감싼 처리를 했었던 경험이 있습니다!

[참고: CLLocationManagerDelegate](https://developer.apple.com/documentation/corelocation/cllocationmanagerdelegate)

> Core Location calls the methods of your delegate object on the runloop from the thread on which you initialized CLLocationManager. That thread must itself have an active run loop, like the one found in your app’s main thread.

mienne reviewed (approved):
수고하셨습니다👍

wnsxor1993 commented:
안녕하세요 두두.
금일 신규 PR 작성을 위해 현재의 PR은 머지하고자 합니다.
말씀주신 사항은 금일 PR에 같이 반영하여 공유드리도록 하겠습니다ㅎㅎ
